assertion failed: position not set for new com.augustnagro.magnum.SqlName # -1 of class dotty.tools.dotc.ast.Trees$Select

Hello!

## Compiler version

`3.4.2`

## Minimized code



<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

With dependency: `com.augustnagro::magnum::1.2.0`

`Schema.scala`:
```Scala
package backend.db

import com.augustnagro.magnum.*
import com.augustnagro.magnum.Spec as MSpec


@Table(SqliteDbType)
@SqlName("effect_event")
case class EffectEvent() derives DbCodec

object EffectEvent {
  case class Creator()
}
```

`Test.scala` (I wasn't able to reproduce with only one file):
```Scala
package backend

import com.augustnagro.magnum
import com.augustnagro.magnum.*

val effectEventTable = TableInfo[db.EffectEvent.Creator, db.EffectEvent, Long]
```

`TableInfo`: https://github.com/AugustNagro/magnum/blob/v1.2.0/magnum/src/main/scala/com/augustnagro/magnum/TableInfo.scala#L22

1. Do a clean compile
2. Change `Test.scala`: `echo "//" >> Test.scala`
3. compile again

Repository with reproduction project: https://github.com/fdietze/scala-bug-magnum-reproduction/tree/main


## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
[info]   exception occurred while typechecking /home/felix/projects/scala-bug-magnum-reproduction/backend/src/Test.scala
[info]
[info]   An unhandled exception was thrown in the compiler.
[info]   Please file a crash report here:
[info]   https://github.com/scala/scala3/issues/new/choose
[info]   For non-enriched exceptions, compile with -Yno-enrich-error-messages.
[info]
[info]      while compiling: /home/felix/projects/scala-bug-magnum-reproduction/backend/src/Test.scala
[info]         during phase: typer
[info]                 mode: Mode(ImplicitsEnabled)
[info]      library version: version 2.13.12
[info]     compiler version: version 3.4.2
[info]             settings: -classpath /home/felix/.cache/coursier/v1/https/repo1.maven.org/maven2/com/augustnagro/magnum_3/1.2.0/magnum_3-1.2.0.jar:/home/felix/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.4.2/scala3-library_3-3.4.2.jar:/home/felix/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar:/home/felix/projects/scala-bug-magnum-reproduction/backend/compile-resources:/home/felix/projects/scala-bug-magnum-reproduction/out/backend/compile.dest/classes -d /home/felix/projects/scala-bug-magnum-reproduction/out/backend/compile.dest/classes
[error] ## Exception when compiling 2 sources to /home/felix/projects/scala-bug-magnum-reproduction/out/backend/compile.dest/classes
[error] java.lang.AssertionError: assertion failed: position not set for new com.augustnagro.magnum.SqlName # -1 of class dotty.tools.dotc.ast.Trees$Select in <no file>
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.typer.Typer$.assertPositioned(Typer.scala:76)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3293)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3413)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:957)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1117)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:351)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedApply(Inliner.scala:819)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3137)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.ReTyper.typedInlined(ReTyper.scala:100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3160)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedSelect(Inliner.scala:773)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3112)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3220)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3295)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.ReTyper.typedInlined(ReTyper.scala:100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3160)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.ReTyper.typedInlined(ReTyper.scala:100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3160)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3413)
[error] dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:2564)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedValDef(Inliner.scala:815)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3116)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3220)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3324)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3370)
[error] dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1209)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1213)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3145)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.ReTyper.typedInlined(ReTyper.scala:100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3160)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3413)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedSplice(Inliner.scala:840)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3182)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:918)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3295)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.inlines.Inliner.inlined(Inliner.scala:680)
[error] dotty.tools.dotc.inlines.Inlines$InlineCall.expand(Inlines.scala:492)
[error] dotty.tools.dotc.inlines.Inlines$.inlineCall(Inlines.scala:163)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgsOther$1(Typer.scala:4072)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:4180)
[error] dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4407)
[error] dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3701)
[error] dotty.tools.dotc.typer.Typer.readapt$1(Typer.scala:3712)
[error] dotty.tools.dotc.typer.Typer.readaptSimplified$1(Typer.scala:3713)
[error] dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:3945)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:3954)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:4156)
[error] dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4407)
[error] dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3701)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3413)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1690)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1680)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1690)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1946)
[error] dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:1946)
[error] dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1954)
[error] dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1972)
[error] dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1973)
[error] dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:1984)
[error] dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1728)
[error] dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1734)
[error] dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:804)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:955)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:831)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:178)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:190)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:192)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:398)
[error] dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:3084)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3109)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3220)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3324)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3370)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2814)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3125)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3129)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3220)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3324)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3370)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2947)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3171)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3221)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3298)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3302)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3413)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:47)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:477)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:53)
[error] dotty.tools.dotc.typer.TyperPhase.$anonfun$4(TyperPhase.scala:99)
[error] scala.collection.Iterator$$anon$6.hasNext(Iterator.scala:479)
[error] scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:583)
[error] scala.collection.immutable.List.prependedAll(List.scala:152)
[error] scala.collection.immutable.List$.from(List.scala:684)
[error] scala.collection.immutable.List$.from(List.scala:681)
[error] scala.collection.IterableOps$WithFilter.map(Iterable.scala:898)
[error] dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:100)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:350)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:360)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:360)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:261)
[error] dotty.tools.dotc.Run.compile(Run.scala:246)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:194)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:249)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:184)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:164)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:164)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:212)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:178)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:176)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:454)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:117)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:265)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:409)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:496)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:396)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:170)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] mill.scalalib.worker.ZincWorkerImpl.compileInternal(ZincWorkerImpl.scala:616)
[error] mill.scalalib.worker.ZincWorkerImpl.$anonfun$compileMixed$1(ZincWorkerImpl.scala:392)
[error] mill.api.FixSizedCache.withCachedValue(FixSizedCache.scala:66)
[error] mill.scalalib.worker.ZincWorkerImpl.withCompilers(ZincWorkerImpl.scala:473)
[error] mill.scalalib.worker.ZincWorkerImpl.compileMixed(ZincWorkerImpl.scala:391)
[error] mill.scalalib.ScalaModule.$anonfun$compile$2(ScalaModule.scala:280)
[error] mill.define.Task$TraverseCtx.evaluate(Task.scala:71)
[error] mill.eval.GroupEvaluator.$anonfun$evaluateGroup$11(GroupEvaluator.scala:356)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$5(SystemStreams.scala:78)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$4(SystemStreams.scala:77)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$3(SystemStreams.scala:76)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withErr(Console.scala:193)
[error] mill.api.SystemStreams$.$anonfun$withStreams$2(SystemStreams.scala:75)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withOut(Console.scala:164)
[error] mill.api.SystemStreams$.$anonfun$withStreams$1(SystemStreams.scala:74)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withIn(Console.scala:227)
[error] mill.api.SystemStreams$.withStreams(SystemStreams.scala:73)
[error] mill.eval.GroupEvaluator.$anonfun$evaluateGroup$8(GroupEvaluator.scala:356)
[error] mill.eval.GroupEvaluator.$anonfun$evaluateGroup$8$adapted(GroupEvaluator.scala:325)
[error] scala.collection.immutable.Vector.foreach(Vector.scala:2124)
[error] mill.eval.GroupEvaluator.computeAll$1(GroupEvaluator.scala:325)
[error] mill.eval.GroupEvaluator.evaluateGroup(GroupEvaluator.scala:380)
[error] mill.eval.GroupEvaluator.$anonfun$evaluateGroupCached$22(GroupEvaluator.scala:247)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.eval.GroupEvaluator.$anonfun$evaluateGroupCached$2(GroupEvaluator.scala:238)
[error] mill.eval.GroupEvaluator$synchronizedEval$.$anonfun$apply$1(GroupEvaluator.scala:65)
[error] scala.util.Using$.resource(Using.scala:296)
[error] mill.eval.GroupEvaluator$synchronizedEval$.apply(GroupEvaluator.scala:64)
[error] mill.eval.GroupEvaluator.evaluateGroupCached(GroupEvaluator.scala:83)
[error] mill.eval.GroupEvaluator.evaluateGroupCached$(GroupEvaluator.scala:71)
[error] mill.eval.EvaluatorImpl.evaluateGroupCached(EvaluatorImpl.scala:15)
[error] mill.eval.EvaluatorCore.$anonfun$evaluate0$2(EvaluatorCore.scala:116)
[error] scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
[error] mill.eval.ExecutionContexts$RunNow$.execute(ExecutionContexts.scala:14)
[error] scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
[error] scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
[error] scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
[error] scala.concurrent.impl.Promise$DefaultPromise.map(Promise.scala:182)
[error] mill.eval.EvaluatorCore.$anonfun$evaluate0$1(EvaluatorCore.scala:92)
[error] mill.eval.EvaluatorCore.$anonfun$evaluate0$1$adapted(EvaluatorCore.scala:90)
[error] scala.collection.immutable.Vector.foreach(Vector.scala:2124)
[error] mill.eval.EvaluatorCore.evaluate0(EvaluatorCore.scala:90)
[error] mill.eval.EvaluatorCore.$anonfun$evaluate$1(EvaluatorCore.scala:43)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.eval.EvaluatorCore.evaluate(EvaluatorCore.scala:34)
[error] mill.eval.EvaluatorCore.evaluate$(EvaluatorCore.scala:26)
[error] mill.eval.EvaluatorImpl.evaluate(EvaluatorImpl.scala:15)
[error] mill.main.RunScript$.evaluateNamed(RunScript.scala:38)
[error] mill.main.RunScript$.$anonfun$evaluateTasksNamed$2(RunScript.scala:26)
[error] scala.util.Either.map(Either.scala:390)
[error] mill.main.RunScript$.evaluateTasksNamed(RunScript.scala:26)
[error] mill.runner.MillBuildBootstrap$.evaluateWithWatches(MillBuildBootstrap.scala:399)
[error] mill.runner.MillBuildBootstrap.$anonfun$processFinalTargets$3(MillBuildBootstrap.scala:308)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.runner.MillBuildBootstrap.processFinalTargets(MillBuildBootstrap.scala:308)
[error] mill.runner.MillBuildBootstrap.evaluateRec(MillBuildBootstrap.scala:196)
[error] mill.runner.MillBuildBootstrap.$anonfun$evaluate$1(MillBuildBootstrap.scala:49)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.runner.MillBuildBootstrap.evaluate(MillBuildBootstrap.scala:48)
[error] mill.runner.MillMain$.$anonfun$main0$6(MillMain.scala:234)
[error] mill.runner.Watching$.watchLoop(Watching.scala:27)
[error] mill.runner.MillMain$.$anonfun$main0$1(MillMain.scala:219)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$5(SystemStreams.scala:78)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$4(SystemStreams.scala:77)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] mill.api.SystemStreams$.$anonfun$withStreams$3(SystemStreams.scala:76)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withErr(Console.scala:193)
[error] mill.api.SystemStreams$.$anonfun$withStreams$2(SystemStreams.scala:75)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withOut(Console.scala:164)
[error] mill.api.SystemStreams$.$anonfun$withStreams$1(SystemStreams.scala:74)
[error] scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error] scala.Console$.withIn(Console.scala:227)
[error] mill.api.SystemStreams$.withStreams(SystemStreams.scala:73)
[error] mill.runner.MillMain$.main0(MillMain.scala:101)
[error] mill.runner.MillServerMain$.main0(MillServerMain.scala:83)
[error] mill.runner.MillServerMain$.main0(MillServerMain.scala:35)
[error] mill.runner.Server.$anonfun$handleRun$1(MillServerMain.scala:187)
[error] java.base/java.lang.Thread.run(Thread.java:1583)
[error]
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -795,17 +795,14 @@ class TreeUnpickler(reader: TastyReader,
       readLater(end, reader =>
         val tp = reader.readType()
         def readAnnotTree(rdr: TreeReader)(using Context) =
+          if isCompactAnnotTypeTag(rdr.reader.nextByte) then TypeTree(rdr.readType())
           // Annotation trees may be inspected by macros via `annot.tree` and
           // spliced into a macro expansion; the re-typer of that expansion
           // requires every untyped tree to have a span (Typer.assertPositioned).
           // For incremental compilation the enclosing unpickling does not use
           // Mode.ReadPositions, so without this the deserialized annotation
           // tree has no spans and the re-typer crashes (issue #21383).
-          // Mirrors the handling of inline method bodies above.
-          val ctx1 = ctx.addMode(Mode.ReadPositions)
-          inContext(rdr.sourceChangeContext(Addr(0))(using ctx1)):
-            if isCompactAnnotTypeTag(rdr.reader.nextByte) then TypeTree(rdr.readType())
-            else rdr.readTree()
+          else rdr.readTree()(using ctx.addMode(Mode.ReadPositions))
         val lazyAnnotTree = reader.readLaterWithOwner(end, readAnnotTree(_))
         owner =>
           new DeferredSymAndTree(tp.typeSymbol, lazyAnnotTree(owner).complete):

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -795,8 +795,17 @@ class TreeUnpickler(reader: TastyReader,
       readLater(end, reader =>
         val tp = reader.readType()
         def readAnnotTree(rdr: TreeReader)(using Context) =
-          if isCompactAnnotTypeTag(rdr.reader.nextByte) then TypeTree(rdr.readType())
-          else rdr.readTree()
+          // Annotation trees may be inspected by macros via `annot.tree`. The
+          // reflected tree must have positions (Typer.assertPositioned), so we
+          // force Mode.ReadPositions even if the enclosing unpickling didn't.
+          // Without this, when TASTY is read for incremental compilation
+          // (which does not use ReadPositions by default), macros see trees
+          // whose internal nodes are missing spans, which causes typer
+          // crashes after re-typing the macro expansion (issue #21383).
+          val ctx1 = ctx.addMode(Mode.ReadPositions)
+          inContext(rdr.sourceChangeContext()(using ctx1)):
+            if isCompactAnnotTypeTag(rdr.reader.nextByte) then TypeTree(rdr.readType())
+            else rdr.readTree()
         val lazyAnnotTree = reader.readLaterWithOwner(end, readAnnotTree(_))
         owner =>
           new DeferredSymAndTree(tp.typeSymbol, lazyAnnotTree(owner).complete):

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -795,15 +795,15 @@ class TreeUnpickler(reader: TastyReader,
       readLater(end, reader =>
         val tp = reader.readType()
         def readAnnotTree(rdr: TreeReader)(using Context) =
-          // Annotation trees may be inspected by macros via `annot.tree`. The
-          // reflected tree must have positions (Typer.assertPositioned), so we
-          // force Mode.ReadPositions even if the enclosing unpickling didn't.
-          // Without this, when TASTY is read for incremental compilation
-          // (which does not use ReadPositions by default), macros see trees
-          // whose internal nodes are missing spans, which causes typer
-          // crashes after re-typing the macro expansion (issue #21383).
+          // Annotation trees may be inspected by macros via `annot.tree` and
+          // spliced into a macro expansion; the re-typer of that expansion
+          // requires every untyped tree to have a span (Typer.assertPositioned).
+          // For incremental compilation the enclosing unpickling does not use
+          // Mode.ReadPositions, so without this the deserialized annotation
+          // tree has no spans and the re-typer crashes (issue #21383).
+          // Mirrors the handling of inline method bodies above.
           val ctx1 = ctx.addMode(Mode.ReadPositions)
-          inContext(rdr.sourceChangeContext()(using ctx1)):
+          inContext(rdr.sourceChangeContext(Addr(0))(using ctx1)):
             if isCompactAnnotTypeTag(rdr.reader.nextByte) then TypeTree(rdr.readType())
             else rdr.readTree()
         val lazyAnnotTree = reader.readLaterWithOwner(end, readAnnotTree(_))

--- a/sbt-test/source-dependencies/i21383/Macro.scala
+++ b/sbt-test/source-dependencies/i21383/Macro.scala
@@ -1,0 +1,14 @@
+import scala.quoted.*
+
+class MyAnnot(val name: String) extends scala.annotation.StaticAnnotation
+
+object Macro:
+  transparent inline def annotOf[T]: String =
+    ${ annotOfImpl[T] }
+
+  def annotOfImpl[T: Type](using Quotes): Expr[String] =
+    import quotes.reflect.*
+    val annTerm = TypeRepr.of[T].typeSymbol.annotations
+      .find(_.tpe =:= TypeRepr.of[MyAnnot]).get
+    val annExpr = annTerm.asExprOf[MyAnnot]
+    '{ $annExpr.name }

--- a/sbt-test/source-dependencies/i21383/Schema.scala
+++ b/sbt-test/source-dependencies/i21383/Schema.scala
@@ -1,0 +1,2 @@
+@MyAnnot("hello")
+class Foo

--- a/sbt-test/source-dependencies/i21383/Test.scala
+++ b/sbt-test/source-dependencies/i21383/Test.scala
@@ -1,0 +1,1 @@
+val x = Macro.annotOf[Foo]

--- a/sbt-test/source-dependencies/i21383/changes/Test.scala
+++ b/sbt-test/source-dependencies/i21383/changes/Test.scala
@@ -1,0 +1,2 @@
+// Edit to trigger incremental recompilation of Test.scala while Schema.tasty is loaded from disk.
+val x = Macro.annotOf[Foo]

--- a/sbt-test/source-dependencies/i21383/project/DottyInjectedPlugin.scala
+++ b/sbt-test/source-dependencies/i21383/project/DottyInjectedPlugin.scala
@@ -1,0 +1,11 @@
+import sbt._
+import Keys._
+
+object DottyInjectedPlugin extends AutoPlugin {
+  override def requires = plugins.JvmPlugin
+  override def trigger = allRequirements
+
+  override val projectSettings = Seq(
+    scalaVersion := sys.props("plugin.scalaVersion")
+  )
+}

--- a/sbt-test/source-dependencies/i21383/test
+++ b/sbt-test/source-dependencies/i21383/test
@@ -1,0 +1,6 @@
+> compile
+# Recompile Test.scala while Schema.tasty is read from disk. Without the fix
+# for issue #21383 the re-typing of the annotation tree returned by the macro
+# asserts on a missing position.
+$ copy-file changes/Test.scala Test.scala
+> compile

--- a/tests/pos-macros/i21383/Macro_1.scala
+++ b/tests/pos-macros/i21383/Macro_1.scala
@@ -1,0 +1,17 @@
+import scala.quoted.*
+
+class MyAnnot(val name: String) extends scala.annotation.StaticAnnotation
+
+@MyAnnot("hello")
+class Foo
+
+object Macro:
+  transparent inline def annotOf[T]: String =
+    ${ annotOfImpl[T] }
+
+  def annotOfImpl[T: Type](using Quotes): Expr[String] =
+    import quotes.reflect.*
+    val annTerm = TypeRepr.of[T].typeSymbol.annotations
+      .find(_.tpe =:= TypeRepr.of[MyAnnot]).get
+    val annExpr = annTerm.asExprOf[MyAnnot]
+    '{ $annExpr.name }

--- a/tests/pos-macros/i21383/Test_2.scala
+++ b/tests/pos-macros/i21383/Test_2.scala
@@ -1,0 +1,1 @@
+val x = Macro.annotOf[Foo]

